pkg/compact/compact.go

```diff
@@ -157,6 +160,42 @@ func (c *Syncer) SyncMetas(ctx context.Context) error {
 }
 
 func (c *Syncer) syncMetas(ctx context.Context) error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	metaIDsChan := make(chan ulid.ULID)
+	errChan := make(chan error, c.metaSyncConcurrency)
+
+	workCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for i := 0; i < c.metaSyncConcurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for id := range metaIDsChan {
+				// Check if we already have this block cached locally.
+				c.blocksMtx.Lock()
+				_, seen := c.blocks[id]
+				c.blocksMtx.Unlock()
```

Comment:
There is a potential race condition here - if for some reason `id` appears many times in `metaIDsChan` then we could make many requests to fetch the metadata, since we don't actually update the local view until we've successfully downloaded.

We could add some extra stuff to stop this race ever happening, but I don't think it's necessarily worth it as I think the bucket iterator works so that we only see the `id` once anyway.  In the worst case, it just means we are doing some unnecessary requests.

Reply:
No it's fine I think. There is no way to have the same `key` in object store listing, so all good.